Parser/Services/HtmlParserService.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using Parser.Infrastructure;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace Parser.Services
+{
+    public class HtmlParserService : IHtmlParserService
+    {
+
+        private readonly IEnumerable<IHtmlProvider> _providers;
+        
+        public HtmlParserService(IEnumerable<IHtmlProvider> providers)
+        {
+            _providers = providers;
+        }
+        
+        public async Task<string> GetSiteTitleAsync(string url)
+        {
+            var provider = await FindImplementation(url);
+            
+            string title = await GetTitle(await provider.GetPageHtmlAsync(url));
+            if (string.IsNullOrEmpty(title))
+            {
+                throw new ArgumentException();
+            }
+            return title;
+        }
+
+        private async Task<string> GetTitle(string html)
+        {
+            return Regex.Match(html, @"\<title\b[^>]*\>\s*(?<Title>[\s\S]*?)\</title\>",
+                RegexOptions.IgnoreCase).Groups["Title"].Value;
+        }
+
+        private Task<IHtmlProvider> FindImplementation(string url)
+        {
+            foreach (var provider in _providers)
+            {
+                if (url.Contains(provider.Domain))
+                {
+                    return Task.FromResult(provider);
+                }
```

Comment:
What's wrong with this implementation? In which cases it may not work? What does it depend on?